File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-2.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) return null;
        HashMap<Node, Integer> map = new HashMap<>();
        List<Node> list = new ArrayList<>();

        Node pom = head;
        int i = 0;
        while(pom != null){
            if(i != 0){
                Node newNode = new Node(pom.val);
                list.add(newNode);
                list.get(i-1).next = newNode;
            }else{
                list.add(new Node(pom.val));
            }
            map.put(pom, i++);
            pom = pom.next;

        }
        pom = head;
        i = 0;
        while(pom != null){

            if(pom.random == null){
                list.get(i++).random = null;
            }else{
                list.get(i++).random=list.get(map.get(pom.random));
            }
            
            pom = pom.next;
        }
        return list.get(0);
    }
}

File: Data Structures & Algorithms/reverse-linked-list-ii/submission-2.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun reverseBetween(head: ListNode?, left: Int, right: Int): ListNode? {
        var subListPrev: ListNode? = if(left > 1) head else null
        for(i in 2 until left){
            subListPrev = subListPrev?.next
        }
        val firstNodeInSub = subListPrev?.next ?: head
        var pom = firstNodeInSub
        var pom2 = firstNodeInSub?.next
        for(i in left until right){
            val copy = pom2
            pom2 = pom2?.next
            copy?.let{
                it.next = pom
            }
            pom = copy
        }
        subListPrev?.let{
            it.next = pom
        }
        firstNodeInSub?.let{
            it.next = pom2
        }
        return if(left > 1) head else pom 
    }
}
